core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -202,139 +206,148 @@ private boolean unsubscribe(String queryName,
+                      .stream()
+                      .filter(filter)
+                      .map(updateHandlers::get)
+                      .filter(Objects::nonNull)
+                      .forEach(FluxSinkWrapper::complete);
+    }
+
+    @Override
+    public void completeExceptionally(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter, Throwable cause) {
+        updateHandlers.keySet()
+                      .stream()
+                      .filter(filter)
+                      .map(updateHandlers::get)
+                      .filter(Objects::nonNull)
+                      .forEach(uh -> uh.error(cause));
     }
 
     /**
-     * Invokes subscription query handler with freshly initialized emitter.
-     *
-     * @param uow           the Unit of Work in which the query handler will be invoked
-     * @param queryHandler  the query handler to be invoked
-     * @param updateHandler the update handler to be invoked with result of query handler
-     * @param <Q>           the query type
-     * @param <I>           the initial result type
-     * @param <U>           the incremental update type
-     * @return handle to cancel updates on this query
+     * Provides the set of running subscription queries. If there are changes to subscriptions they will be reflected in
+     * the returned set of this method.
      *
-     * @throws Exception propagated from query handler
+     * @return the set of running subscription queries
      */
-    private <Q, I, U> Registration invokeSubscriptionQueryHandler(UnitOfWork<QueryMessage<Q, I>> uow,
-                                                                  SubscriptionQueryMessageHandler<? super QueryMessage<?, I>, I, U> queryHandler,
-                                                                  UpdateHandler<I, U> updateHandler)
-            throws Exception {
-        SimpleQueryUpdateEmitter<I, U> emitter = new SimpleQueryUpdateEmitter<>(updateHandler);
-        I initialResult = interceptAndInvoke(uow, m -> queryHandler.handle(m, emitter)).getPayload();
-        emitter.initialResult(initialResult);
-        return () -> {
-            emitter.cancelRegistration();
-            return true;
-        };
+    public Set<SubscriptionQueryMessage<?, ?, ?>> activeSubscriptions() {
+        return Collections.unmodifiableSet(updateHandlers.keySet());
+    }
+
+    @SuppressWarnings("unchecked")
+    private <U> void doEmit(SubscriptionQueryMessage<?, ?, ?> query, FluxSinkWrapper<?> updateHandler,
+                            SubscriptionQueryUpdateMessage<U> update) {
+        MessageMonitor.MonitorCallback monitorCallback = updateMessageMonitor.onMessageIngested(update);
+        try {
+            ((FluxSinkWrapper<SubscriptionQueryUpdateMessage<U>>) updateHandler).next(update);
+            monitorCallback.reportSuccess();
+        } catch (Exception e) {
+            logger.error(format(
+                    "An error happened while trying to emit an update to a query: %s.",
```

Comment:
Might be nice to inform people that the update handler will be removed entirely, so that they know that they cant expect any further updates after this.